airflow/models/dag.py

```diff
@@ -1025,11 +1026,110 @@ def get_task_instances(self, start_date=None, end_date=None, state=None, session
             start_date = (timezone.utcnow() - timedelta(30)).date()
             start_date = timezone.make_aware(datetime.combine(start_date, datetime.min.time()))
 
-        tis = session.query(TaskInstance).filter(
-            TaskInstance.dag_id == self.dag_id,
-            TaskInstance.execution_date >= start_date,
-            TaskInstance.task_id.in_([t.task_id for t in self.tasks]),
+        return (
+            self._get_task_instances(
+                task_ids=None,
+                start_date=start_date,
+                end_date=end_date,
+                state=state,
+                include_subdags=False,
+                include_parentdag=False,
+                include_dependent_dags=False,
+                as_pk_tuple=False,
+                session=session,
+            )
+            .order_by(TaskInstance.execution_date)
+            .all()
         )
+
+    @overload
+    def _get_task_instances(
+        self,
+        *,
+        task_ids,
+        start_date: Optional[datetime],
+        end_date: Optional[datetime],
+        state: Union[str, List[str]],
+        include_subdags: bool,
+        include_parentdag: bool,
+        include_dependent_dags: bool,
+        as_pk_tuple: Literal[True],
+        session: Session,
+        dag_bag: "DagBag" = None,
+        recursion_depth: int = 0,
+        max_recursion_depth: int = None,
+        visited_external_tis: Set[Tuple[str, str, datetime]] = None,
+    ) -> Set[Tuple[str, str, datetime]]:
+        ...
+
+    @overload
+    def _get_task_instances(
+        self,
+        *,
+        task_ids,
+        start_date: Optional[datetime],
+        end_date: Optional[datetime],
+        state: Union[str, List[str]],
+        include_subdags: bool,
+        include_parentdag: bool,
+        include_dependent_dags: bool,
+        as_pk_tuple: Literal[False],
+        session: Session,
+        dag_bag: "DagBag" = None,
+        recursion_depth: int = 0,
+        max_recursion_depth: int = None,
+        visited_external_tis: Set[Tuple[str, str, datetime]] = None,
+    ) -> Iterable[TaskInstance]:
+        ...
```

Comment:
I added a `typing.overload` here so that mypy etc knows the right type based on the `as_pk_tuple` value.